File: codegen-gradle-plugin/src/test/java/io.vrap.rmf.codegen.common/BuildLogicFunctionalTest.java
//package io.vrap.rmf.codegen.common;
//
//import org.assertj.core.api.Assertions;
//import org.gradle.testkit.runner.BuildResult;
//import org.gradle.testkit.runner.GradleRunner;
//import org.gradle.testkit.runner.TaskOutcome;
//import org.junit.Before;
//import org.junit.Rule;
//import org.junit.Test;
//import org.junit.rules.TemporaryFolder;
//import org.slf4j.Logger;
//import org.slf4j.LoggerFactory;
//
//import java.io.BufferedWriter;
//import java.io.File;
//import java.io.FileWriter;
//import java.io.IOException;
//import java.nio.file.Paths;
//
//public class BuildLogicFunctionalTest {
//
//    @Rule
//    public final TemporaryFolder testProjectDir = new TemporaryFolder();
//    private File buildFile;
//
//    private final Logger logger = LoggerFactory.getLogger(getClass());
//
//    @Before
//    public void setup() throws IOException {
//        buildFile = testProjectDir.newFile("build.gradle");
//    }
//
//    @Test
//    public void testGenerateEntitiesTask() throws IOException {
//
//        final File apiFile = Paths.get("src/test/resources/api-spec/api.raml").toAbsolutePath().toFile();
//        System.err.println(apiFile);
//        Assertions.assertThat(apiFile).exists();
//
//        String buildFileContent =
//                        "plugins {\n" +
//                        "    id 'io.vrap.rmf.codegen-plugin'\n" +
//                        "}\n" +
//                        "generateRamlStub{\n" +
//                        "    ramlFileLocation= file('"+apiFile.getAbsolutePath()+"')\n" +
//                        "    outputFolder=file('"+testProjectDir.getRoot().toString()+"')\n" +
//                        "    packagePrefix= 'com.commercetools'\n" +
//                        "}";
//
//        writeFile(buildFile, buildFileContent);
//
//        BuildResult result = GradleRunner.create()
//                .withProjectDir(testProjectDir.getRoot())
//                .withArguments("generateRamlStub","--stacktrace")
//                .withPluginClasspath()
//                .build();
//
//        logger.warn(result.getOutput());
//
//        Assertions.assertThat(result.task(":generateRamlStub").getOutcome()).isEqualTo(TaskOutcome.SUCCESS);
//
//        Assertions.assertThat(testProjectDir.getRoot().list((dir, name) -> name.equals("com"))).isNotEmpty();
//
//
//    }
//
//    private void writeFile(File destination, String content) throws IOException {
//        BufferedWriter output = null;
//        try {
//            output = new BufferedWriter(new FileWriter(destination));
//            output.write(content);
//        } finally {
//            if (output != null) {
//                output.close();
//            }
//        }
//    }
//}

File: codegen-gradle-plugin/src/test/kotlin/io.vrap.rmf.codegen.common/BuildLogicFunctionalTest.kt
package io.vrap.rmf.codegen.common

import org.assertj.core.api.Assertions
import org.gradle.testkit.runner.GradleRunner
import org.gradle.testkit.runner.TaskOutcome
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TemporaryFolder
import org.slf4j.LoggerFactory
import java.io.BufferedWriter
import java.io.File
import java.io.FileWriter
import java.nio.file.Paths

class BuildLogicFunctionalTest {

    @get:Rule
    val testProjectDir = TemporaryFolder()

    private lateinit var buildFile: File
    private val logger = LoggerFactory.getLogger(javaClass)

    @Before
    fun setup() {
        buildFile = testProjectDir.newFile("build.gradle")
    }

    @Test
    fun testGenerateEntitiesTask() {

        val apiFile = Paths.get("../api-spec/api.raml").toAbsolutePath().toFile()
        Assertions.assertThat(apiFile).exists()

        val buildFileContent = """
        |plugins {
        |    id 'io.vrap.rmf.codegen-plugin'
        |}
        |
        |RamlGenerator {
        |    apiSpec {
        |      uri = file('$apiFile')
        |      targets {
        |          importApi {
        |               path = file('import-api/build/generated-classes')
        |               models_package = 'com.commercetools.models'
        |               client_package = 'com.commercetools.client'
        |               target = 'typescriptModel'
        |          }
        |          importApi_groovy_dsl {
        |               path = file('import-api/build/generated-classes')
        |               models_package = 'com.commercetools.models'
        |               client_package = 'com.commercetools.client'
        |               target = 'groovyDsl'
        |          }
        |          importStorageService {
        |               path = file('import-api2/build/generated-classes')
        |               models_package = 'com.commercetools.models2'
        |               client_package = 'com.commercetools.client2'
        |               target = 'javaModel'
        |          }
        |          importStorageService {
        |               path = file('import-api2/build/generated-classes')
        |               models_package = 'com.commercetools.models2'
        |               client_package = 'com.commercetools.client2'
        |               target = 'joiValidator'
        |          }
        |          springClient {
        |               path = file('import-api3/build/generated-classes')
        |               models_package = 'com.commercetools.models2'
        |               client_package = 'com.commercetools.client2'
        |               target = 'javaSpringClient'
        |          }
        |       }
        |    }
        |}
        """.trimMargin()

        writeFile(buildFile, buildFileContent)
        val result = GradleRunner.create()
                .withProjectDir(testProjectDir.root)
                .withArguments("generateRamlStub", "--stacktrace")
                .withPluginClasspath()
                .build()

        logger.warn(result.output)

        Assertions.assertThat(result.task(":generateRamlStub")!!.outcome).isEqualTo(TaskOutcome.SUCCESS)

        Assertions.assertThat(testProjectDir.root.list { _, name -> name == "import-api" }).isNotEmpty
    }

    private fun writeFile(destination: File?, content: String) {
        var output: BufferedWriter? = null
        try {
            output = BufferedWriter(FileWriter(destination!!))
            output.write(content)
        } finally {
            output?.close()
        }
    }
}
